[DOCS] Updates the "Interactive Mode" guide for creating Expectations (#7624)
great-expectations/great_expectations: docs/docusaurus/docs/guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data.md

---
title: How to create Expectations interactively in Python
tag: [how-to, getting started]

> description: A how-to guide describing how to create Expectations in a Python interpreter or script while interactively receiving feedback by validating them against a Batch of data.

Reviewer: ```suggestion
description: Create Expectations with a Python interpreter or a script and then use interactive feedback to validate them with batch data.
```

keywords: [Expectations, Interactive Mode, Interactive]
---

import Prerequisites from '/docs/components/_prerequisites.jsx'
import PrerequisiteQuickstartGuideComplete from '/docs/components/prerequisites/_quickstart_completed.mdx'
import IfYouStillNeedToSetupGX from '/docs/components/prerequisites/_if_you_still_need_to_setup_gx.md'
import DataContextInitializeQuickOrFilesystem from '/docs/components/setup/link_lists/_data_context_initialize_quick_or_filesystem.mdx'
import ConnectingToDataFluently from '/docs/components/connect_to_data/link_lists/_connecting_to_data_fluently.md'

> In order to Validate data, we must first define a set of Expectations for that data to be validated against.  In this guide, we will walk you through processes of interactively creating and editing Expectations while simultaneously validating them against a Batch of data. By validating our Expectations as we define them we can quickly see if the Expectations we are setting are suitable for our data, and adjust them if necessary.

Reviewer: ```suggestion
To validate data, you need to first define the Expectations that you'll use for your data validation.  In this guide, you'll learn how to interactively create and edit Expectations and then validate them with batch data. Validating your Expectations as you define them allows you to quickly determine if the Expectations are suitable for your data, and identify where changes might be necessary.
```

> :::tip Does this process edit my data?

Reviewer: I would argue that this information shouldn't be in a tip admonition. It's not a tip. It's simply informative. Punctuation in headings should be used sparingly, if at all.

> Not at all.  We call this process the "Interactive method" of creating and editing Expectations because validating the Expectations as we work requires interacting with a Batch of data.  This data is introspected to see if it fits the defined Expectation, but it is not edited: only examined.

Reviewer: ```suggestion
No.  The interactive method used to create and edit Expectations does not edit or alter the batch data.
```
If you don't accept the edit, I recommend replacing _introspected_ with a simpler term.

:::

## Prerequisites

<Prerequisites>

- Great Expectations installed in a Python environment

> - Initialized a Filesystem Data Context to save your Expectations in

Reviewer: ```suggestion
- A Filesystem Data Context for your Expectations
```
If you don't accept the edit, I recommend finding a simpler term than _Initialized_

> - Created a Datasource from which to request a Batch of data for introspection

Reviewer: ```suggestion
- A Datasource that you can use to request and review batch data
```
If you don't accept the edit, I recommend replacing _introspection_ with a simpler term.

> - A passion for data quality

Reviewer: ```suggestion
- A passion for data quality
```
I would remove this

</Prerequisites> 

<details>
<summary>

> ### If you still need to set up Great Expectations...

Reviewer: ```suggestion
### If you haven't set up Great Expectations
```

</summary>

<IfYouStillNeedToSetupGX />

</details>

<details>
<summary>

> ### If you still need to initialize your Data Context...

Reviewer: ```suggestion
### If you haven't initialized your Data Context
```

</summary>

> Please see the appropriate guide from the following:

Reviewer: ```suggestion
Select one of the following guides:
```

<DataContextInitializeQuickOrFilesystem />

</details>

<details>
<summary>

### If you still need to create a Datasource...

</summary>

> Please reference the appropriate guide from the following:

Reviewer: ```suggestion
Select one of the following guides:
```
Is there an inappropriate guide? :)

![image](https://user-images.githubusercontent.com/104205257/231878510-57d9f538-739c-4867-97e7-2d1ef6b554b2.png)

Author: The inappropriate guides would be the guides that correspond to environments or source data systems that don't match the one they're using. :P

<ConnectingToDataFluently />

</details>

## Steps

### 1. Import the Great Expectations module and instantiate a Data Context

> For this guide we will be working with Python code in a Jupyter Notebook. Jupyter is included with GX and provides a very convenient interface that lets us easily edit code and immediately see the result of our changes.

Reviewer: ```suggestion
To complete the procedures in this guide, you'll use Python code in a Jupyter Notebook. Jupyter is included with GX and it lets you edit code and immediately see the results of your changes.
```
What are users doing here? Tell them what they'll accomplish - _You'll use Python code in a Jupyter Notebook to..._

![image](https://user-images.githubusercontent.com/104205257/231879477-9680f7b3-d0b8-488b-abf0-38c372296fec.png)

> The code to import Great Expectations and instantiate a Data Context is:

Reviewer: ```suggestion
Run the following code to import Great Expectations and instantiate a Data Context:
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py imports and data context"
```

:::info Data Contexts and persisting data

> If you are using an Ephemeral Data Context, your configurations will not persist beyond the current Python session.  If you are using a Filesystem or Cloud Data Context, they will.  The `get_context()` method will return the first of a Cloud or Filesystem Data Context if it can find one, but if it can't then it will provide you with an Ephemeral Data Context.  For more information on the `get_context()` method, please see our guide on [how to quickly instantiate a Data Context](/docs/guides/setup/configuring_data_contexts/instantiating_data_contexts/how_to_quickly_instantiate_a_data_context).

Reviewer: ```suggestion
If you're using an Ephemeral Data Context, your configurations won't persist beyond the current Python session.  However, if you're using a Filesystem or Cloud Data Context, they do persist.  The `get_context()` method returns the first Cloud or Filesystem Data Context it can find. If a Cloud or Filesystem Data Context can't be located, it provides an Ephemeral Data Context. For more information about the `get_context()` method, see [How to quickly instantiate a Data Context](/docs/guides/setup/configuring_data_contexts/instantiating_data_contexts/how_to_quickly_instantiate_a_data_context).
```

:::

### 2. Use an existing Data Asset to create a Batch Request

> From the Data Context we initialized we can access our previously created Datasources and Data Assets.  We will retrieve a previously configured Data Asset, and then create a Batch Request indicating the Batch of data that we intend to Validate our Expectations against:

Reviewer: ```suggestion
Add the following method to retrieve a previously configured Data Asset from the Data Context you initialized and create a Batch Request to identify the batch data that you'll use to validate your Expectations:
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py get_data_asset_and_build_batch_request"
```

> :::info Limiting the Batches returned by a Batch Request

Reviewer: ```suggestion
:::info Limit the Batches returned by a Batch Request
```

> A Batch Request allows you to limit the Batches returned from a Data Asset by providing a dictionary as the `options` parameter.  If nothing is provided as the `options` parameter, your Batch Request will include all the Batches configured in the Data Asset.  To learn more, please see our guide on [how to request data from a Data Asset](/docs/guides/connecting_to_your_data/fluent/batch_requests/how_to_request_data_from_a_data_asset).

Reviewer: ```suggestion
You can provide a dictionary as the `options` parameter to limit the Batches returned from a Data Asset.  If you leave the `options` parameter empty, your Batch Request includes all of the Batches configured in the Data Asset.  For more information about Batch Requests, see [How to request data from a Data Asset](/docs/guides/connecting_to_your_data/fluent/batch_requests/how_to_request_data_from_a_data_asset).
```

:::

### 3. Create a Validator

> We will use a Validator to interactively create our Expectations.  To do this, a Validator needs two parameters: one will indicate the data to run Expectations against, and the other will provide a name for the combined list of Expectations should we decide to save them for future use.

Reviewer: ```suggestion
To use a Validator to interactively create your Expectations, the Validator needs two parameters. One parameter indicates the batch data that is used to validate the Expectations. The other parameter provides a name for the combined list of Expectations that you can save for future use.
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py get_validator_and_inspect_data"
```

:::info Working outside a Jupyter Notebook

> In a Jupyter Notebook you will automatically see the results of the above code in a new cell when the code is run.  However, if you are working with a different Python interpreter you may need to explicitly print your results:

Reviewer: ```suggestion
If you're using a Jupyter Notebook you'll automatically see the results of the code you run in a new cell when you run the code. If you're using a different interpreter, you might need to add the following method to explicitly print these results:
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py inspect_data_no_jupyter"
```

:::

### 4. Use the Validator to create and run an Expectation

> The Validator provides access to all of our available Expectations as methods.  By running an `expect_*()` method from the Validator, two things will happen: The Validator will add the specified Expectation to an Expectation Suite (or edit an existing Expectation in the Expectation Suite, if applicable) in its configuration, and the specified Expectation will be run against the data we provided to the Validator with our Batch Request.

Reviewer: ```suggestion
The Validator provides access to all of the available Expectations as methods.  When you run an `expect_*()` method from the Validator, the Validator adds the specified Expectation to an Expectation Suite (or edits an existing Expectation in the Expectation Suite, if applicable) in its configuration, and then the specified Expectation is run against the data you provided to the Validator with your Batch Request.
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py interactive_validation"
```

> Since we are working in a Jupyter Notebook, the results of the Validation will be printed after we run an `expect_*()` method.  We can examine those results to determine if the Expectation needs to be edited (by running it again and providing additional optional parameters, or different parameter values).

Reviewer: ```suggestion
If you're using a Jupyter Notebook, the results of the Validation are printed after you run an `expect_*()` method.  You can examine the results to determine if the Expectation needs to be edited. To edit an Expectation, you run it again with additional optional parameters, or different parameter values.
```

:::info Working outside a Jupyter Notebook

> As before, if you are not working in a Jupyter Notebook you may need to explicitly print your results:

Reviewer: ```suggestion
If you're using a different interpreter, you might need to add the following method to explicitly print these results:
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py interactive_validation_no_jupyter"
```

:::

> ### 5. (Optional) Repeat step 4 to create additional Expectations or edit existing ones

Reviewer: ```suggestion
### 5. (Optional) Repeat step 4 to create additional Expectations or edit Expectations
```

> After reviewing the Validation Results that were returned when we created an Expectation, we can choose to edit that Expectation by running the `validator.expect_*()` method with different parameters than previously.  We can also have the Validator run an entirely different `expect_*()` method and create additional Expectations.  All the Expectations that we create will be stored in a list in the Validator's configuration.

Reviewer: ```suggestion
After you've reviewed the Validation Results that were returned when you created an Expectation, you can run the `validator.expect_*()` method with different parameters to edit the Expectation.  You can also have the Validator run an entirely different `expect_*()` method to create additional Expectations.  All the Expectations that you create are stored in a list in the Validator configuration file.
```

:::tip What if I want to use the same Expectation more than once?

> GX takes into account certain parameters when determining if an Expectation is being added to the list or if an existing Expectation should be edited.  For example, if you are using an expectation such as `expect_column_*()` you would edit an existing Expectation by providing the same `column` parameter when repeating the previous steps, and different values for any other parameters.  If you used a different value for the `column` parameter, you will create an additional instance of the Expectation for the new `column` value, rather than overwrite the Expectation you defined with the first `column` value.

Reviewer: ```suggestion
GX takes into account certain parameters when determining if an Expectation is being added to the list or if an existing Expectation should be edited.  For example, if you are using an expectation such as `expect_column_*()` you would edit an existing Expectation by providing the same `column` parameter when repeating the previous steps, and different values for any other parameters.  If you used a different value for the `column` parameter, you will create an additional instance of the Expectation for the new `column` value, rather than overwrite the Expectation you defined with the first `column` value.
```
 _when repeating the previous steps_ - What specific steps?

I think this needs some simplification and clarification

:::

### 6. Save your Expectations for future use

> The Expectations we have created so far are saved in an Expectation Suite on the Validator object.  Since Validators do not persist outside the current Python session these Expectations will not be kept if we stop now.  This can be ideal if we are using a Validator to do some quick data discovery, but in most cases we will want our newly created Expectation Suite to be available in the future.

Reviewer: ```suggestion
The Expectations you created are saved in an Expectation Suite on the Validator object.  Validators do not persist outside the current Python session and for this reason the Expectations are not kept unless **<you do what>**?  This can be ideal if you are using a Validator for quick data validation, but in most cases you'll want to reuse your newly created Expectation Suite.
```

> To keep our Expectations for future use, we will save them to our Data Context.  A Filesystem or Cloud Data Context will persist outside the current Python session, so saving the Expectation Suite in our Data Context's Expectations Store will ensure we can access it in the future:

Reviewer: ```suggestion
To keep your Expectations for future use, you save them to your Data Context.  A Filesystem or Cloud Data Context persists outside the current Python session, so saving the Expectation Suite in your Data Context's Expectations Store ensures you can access it in the future:
```

```python name="tests/integration/docusaurus/validation/validator/how_to_create_and_edit_expectations_with_instant_feedback_fluent.py save_expectation_suite"
```

:::caution Ephemeral Data Contexts and persistence

> Ephemeral Data Contexts do not persist beyond the current Python session.  If you are working with an Ephemeral Data Context, you will need to convert it to a Filesystem Data Context using the Data Context's `convert_to_file_context()` method.  Otherwise, your saved configurations will not be available in future Python sessions as the Data Context itself will no longer be available.

Reviewer: ```suggestion
Ephemeral Data Contexts don't persist beyond the current Python session.  If you're working with an Ephemeral Data Context, you'll need to convert it to a Filesystem Data Context using the Data Context's `convert_to_file_context()` method.  Otherwise, your saved configurations won't be available in future Python sessions as the Data Context itself is no longer available.
```

:::

## Next steps

> Now that you have created and saved an Expectation Suite, you are ready to begin [Validating your data](/docs/guides/validation/validate_data_overview).

Reviewer: ```suggestion
Now that you have created and saved an Expectation Suite, you can [Validate your data](/docs/guides/validation/validate_data_overview).
```
